pennylane/ops/op_math/symbolicop.py

```diff
@@ -263,7 +263,7 @@ def matrix(self, wire_order=None):
             base_matrix = qml.math.convert_like(base_matrix, self.scalar)
         elif scalar_interface == "tensorflow" and not qml.math.iscomplex(self.scalar):
             # cast scalar to complex to avoid an error
-            self.scalar = qml.math.cast_like(self.scalar, base_matrix)
+            self.scalar = qml.math.cast(self.scalar, "complex128")
```

Comment:
this caused a horrific bug when the base op was something like PauliZ, because it casted the scalar to `int` to match the base matrix. Then it tried to reduce a bunch of matrices of zeros and it got ugly

Reply:
Should we instead cast the matrix to match the scalar?

Reply:
Also, can you provide a minimal example for the bug?

Reply:
the failure comes from [this test](https://github.com/PennyLaneAI/pennylane/blob/e76aa4b9ffb65500d21a0da6312455d633327215/tests/interfaces/default_qubit_2_integration/test_tensorflow_default_qubit_2.py#L732). While computing `Sum.diagonalizing_gates`, it tries to reduce some operand matrices. I added a print statement to show which matrices are being reduced, and here is a minimal example:
```pycon
>>> op = tf.Variable(0.1) * qml.PauliZ(0) + tf.Variable(0.2) * qml.PauliY(0)
>>> op.diagonalizing_gates()
reducing matrices:
 tf.Tensor(
[[0 0]
 [0 0]], shape=(2, 2), dtype=int64) 
 tf.Tensor(
[[0.+0.j  0.-0.2j]
 [0.+0.2j 0.+0.j ]], shape=(2, 2), dtype=complex128)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/ops/op_math/composite.py", line 275, in diagonalizing_gates
    eigvecs = tmp_sum.eigendecomposition["eigvec"]
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/ops/op_math/composite.py", line 238, in eigendecomposition
    mat = self.matrix()
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/ops/op_math/sum.py", line 225, in matrix
    reduced_mat, sum_wires = math.reduce_matrices(gen, reduce_func=math.add)
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/math/matrix_manipulation.py", line 307, in reduce_matrices
    reduced_mat, final_wires = reduce(expand_and_reduce, mats_and_wires_gen)
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/math/matrix_manipulation.py", line 305, in expand_and_reduce
    return reduce_func(mat1, mat2), expanded_wires
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/math/multi_dispatch.py", line 151, in wrapper
    return fn(*args, **kwargs)
  File "/Users/matthews/src/github.com/PennyLaneAI/pennylane/pennylane/math/multi_dispatch.py", line 773, in add
    return np.add(*args, **kwargs)
  File "/Users/matthews/.pyenv/versions/pl/lib/python3.9/site-packages/autoray/autoray.py", line 79, in do
    return get_lib_fn(backend, fn)(*args, **kwargs)
  File "/Users/matthews/.pyenv/versions/pl/lib/python3.9/site-packages/tensorflow/python/util/traceback_utils.py", line 153, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/Users/matthews/.pyenv/versions/pl/lib/python3.9/site-packages/tensorflow/python/framework/ops.py", line 7262, in raise_from_not_ok_status
    raise core._status_to_exception(e) from None  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute AddV2 as input #1(zero-based) was expected to be a int64 tensor but is a complex128 tensor [Op:AddV2]
```
re: casting, same as above, "complex128" seems to be what we cast to anyway.